File: src/test/java/edu/metrostate/trackingsystem/infrastructure/utils/PriceDeserializerTest.java
package edu.metrostate.trackingsystem.infrastructure.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonParseException;
import edu.metrostate.trackingsystem.domain.models.Price;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class PriceDeserializerTest {

    private final Gson gson = new GsonBuilder()
            .registerTypeAdapter(Price.class, new PriceDeserializer())
            .create();

    @Test
    public void deserializesNumberOnlyJson() {
        String json = "19.99";
        Price price = gson.fromJson(json, Price.class);
        assertEquals(19.99, price.getPrice(), 0.001);
        assertEquals("dollars", price.getCurrency());
    }

    @Test
    public void deserializesObjectWithPriceAndUnit() {
        String json = "{\"price\": 25.50, \"unit\": \"euros\"}";
        Price price = gson.fromJson(json, Price.class);
        assertEquals(25.50, price.getPrice(), 0.001);
        assertEquals("euros", price.getCurrency());
    }

    @Test
    public void deserializesObjectMissingUnit() {
        String json = "{\"price\": 10.00}";
        Price price = gson.fromJson(json, Price.class);
        assertEquals(10.00, price.getPrice(), 0.001);
        assertEquals("dollars", price.getCurrency());
    }

    @Test
    public void throwsOnInvalidFormat() {
        String json = "\"invalid\"";
        assertThrows(JsonParseException.class, () -> {
            gson.fromJson(json, Price.class);
        });
    }
}

File: src/test/kotlin/edu/metrostate/dealership/infrastructure/utils/PriceDeserializerTest.kt
package edu.metrostate.dealership.infrastructure.utils

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonParseException
import edu.metrostate.dealership.domain.models.Price
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class PriceDeserializerTest {
    private val gson: Gson = GsonBuilder()
        .registerTypeAdapter(Price::class.java, PriceDeserializer())
        .create()

    @Test
    fun deserializesNumberOnlyJson() {
        val json = "19.99"
        val price = gson.fromJson(json, Price::class.java)
        Assertions.assertEquals(19.99, price.price, 0.001)
        Assertions.assertEquals("dollars", price.currency)
    }

    @Test
    fun deserializesObjectWithPriceAndUnit() {
        val json = "{\"price\": 25.50, \"unit\": \"euros\"}"
        val price = gson.fromJson(json, Price::class.java)
        Assertions.assertEquals(25.50, price.price, 0.001)
        Assertions.assertEquals("euros", price.currency)
    }

    @Test
    fun deserializesObjectMissingUnit() {
        val json = "{\"price\": 10.00}"
        val price = gson.fromJson(json, Price::class.java)
        Assertions.assertEquals(10.00, price.price, 0.001)
        Assertions.assertEquals("dollars", price.currency)
    }

    @Test
    fun throwsOnInvalidFormat() {
        val json = "\"invalid\""
        Assertions.assertThrows(
            JsonParseException::class.java
        ) {
            gson.fromJson(json, Price::class.java)
        }
    }
}